### PR TITLE
feat: KnowledgeBase Enhancements + Knowledge Ingestion flow +  Polymorphic Job Tracking

### PR DESCRIPTION
**Pull request by deon-sanchez**
LE-207
### Frontend Work:
@deon-sanchez  

* **New Features**
  * Enabled Knowledge Bases feature
  * Create Knowledge action now displays File and Folder options via dropdown menu
  * Knowledge bases table redesigned with reorganized columns: Source (now sortable), Type, Owner, and Avg Chunk Size
  * Added Status column to knowledge bases display

### Testing Steps for QA
[KNOWLEDGE BASES FEATURE — QA TEST PLAN.pdf](https://github.com/user-attachments/files/25424687/KNOWLEDGE.BASES.FEATURE.QA.TEST.PLAN.pdf)

 
 

### Backend Work:
@dkaushik94 
## Description
This PR introduces a significant refactor of the Knowledge Bases infrastructure alongside generalizations to our Job/Task service and backend safety protocols. The implementations resolve existing data contention issues while enabling cleaner background task handling and paginated results.

### 🚀 Features & Enhancements

* **Knowledge Base Architecture & Storage Refactor**
  * **Class-based Helpers**: Abstracted standalone functions into isolated, responsibility-driven classes ([KBStorageHelper](cci:2://file:///Users/debojitkaushik/ibm/langflow/src/backend/base/langflow/api/utils/kb_helpers.py:36:0-94:87), [KBAnalysisHelper](cci:2://file:///Users/debojitkaushik/ibm/langflow/src/backend/base/langflow/api/utils/kb_helpers.py:97:0-288:44), [KBIngestionHelper](cci:2://file:///Users/debojitkaushik/ibm/langflow/src/backend/base/langflow/api/utils/kb_helpers.py:291:0-449:48)) to significantly improve code modularity and testability.
  * **Chunk Pagination & Filter APIs**: Integrated pagination frameworks and text-match filtering into [get_chunks](cci:1://file:///Users/debojitkaushik/ibm/langflow/src/backend/tests/unit/test_knowledge_bases_api.py:368:4-409:103) endpoints to streamline handling large vector retrieval natively.
  * **Asynchronous Ingestion**: Moved core file data chunking and embedding logic into non-blocking `TaskService` deployments with tracking managed by internal `JobService`. Supports both status polling and dynamic job cancellation.
  * **Chroma/SQLite Contention Fixes**: Addressed historical `Read-Only` exceptions and lock contention by generating a forced fresh Chroma persistent client prior to active storage allocation and ensuring resources are properly garbage-collected upon teardowns.
  * **Improved Metadata Tracking**: Enforces proper tracking of data types like `avg_chunk_size`, `words`, and `source_types`, effectively caching them locally to speed up `/knowledge_bases/` fetching loops by drastically omitting repetitive folder iterations.

* **Job and Task Service Generalizations**
  * Modified background jobs schemas via Alembic migrations. Included `job_type`, `asset_id`, and `asset_type` columns to allow polymorphism. This grants jobs the elasticity to process Canvas evaluations, Datasets mapping, and KB ingestions agnostically without blocking primary router loops.

### 🛠️ Refactoring & Code Quality

* **Security & Error Handling**: Refactored internal `HTTPExceptions`. Raw exception objects are no longer leaked directly to users; issues are safely logged locally (`logger.aerror()`) and users only receive controlled standard text formats globally, addressing blind exception catching.
* **Format Alignments**: Stripped inline `f-string` formats strictly to comply with ruff format `UP031` and corrected file-length guidelines.
* **Testing Expansion**: Augmented `test_knowledge_bases_api` vastly to fully test edge-case states (zero-file uploads, missing dimensions, missing metadata recovery, ingestion rollback). 

### 🐛 Additional Fixes
* Fixed `_get_text_columns` fallback bypass which previously forced `avg_chunk_size` calculations to universally output 0.0. 
* Refreshed base starter project structure files.

### CODE REVIEW

File: src/backend/base/langflow/alembic/versions/169b35510b37_added_job_type_to_make_jobs_generic_and_.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    # Check which columns already exist (handles fresh DB where model creates them)
+    inspector = sa.inspect(conn)
+    existing_columns = {col["name"] for col in inspector.get_columns("job")}
+    existing_indexes = {idx["name"] for idx in inspector.get_indexes("job")}
+
+    with op.batch_alter_table("job", schema=None) as batch_op:
+        if "type" not in existing_columns:
+            batch_op.add_column(
+                sa.Column("type", sa.Enum("workflow", "ingestion", "evaluation", name="job_type_enum"), nullable=True)
+            )
```
Comment (ogabrielluiz) [inline src/backend/base/langflow/alembic/versions/169b35510b37_added_job_type_to_make_jobs_generic_and_.py]: @deon-sanchez Should we do this same check on this op?
https://github.com/langflow-ai/langflow/blob/38082c1bf312a1c699cd7704d251671cb9bb618e/src/backend/base/langflow/alembic/versions/f3b2d1f1002d_add_column_access_type_to_flow.py#L26

Comment (deon-sanchez) [reply src/backend/base/langflow/alembic/versions/169b35510b37_added_job_type_to_make_jobs_generic_and_.py]: @dkaushik94 ^^^

File: src/frontend/src/components/core/dropdownComponent/index.tsx [hunk 2]
```diff
@@ -601,17 +611,39 @@ export default function Dropdown({
...
+              }}
+              onSubmit={(data) => {
+                setOpenDialog(false);
+                setOpen(false);
+                setPendingSelect(data.sourceName);
+                handleRefreshButtonPress();
+              }}
+              hideAdvanced
+            />
+          ) : (
```
Comment (olayinkaadelakun) [inline src/frontend/src/components/core/dropdownComponent/index.tsx:631]: Nit: Due to priority please skip

Responsibility creep: a generic dropdown now “knows” about a specific domain feature (Knowledge Bases). Over time you’ll accumulate more special-cases and the dropdown becomes a god-component.

Harder to test safely: changes for Knowledge Base flows can accidentally regress unrelated dropdown usages.

Global dependency + bundle weight: importing KnowledgeBaseUploadModal into a core component increases the surface area and load cost for every place the dropdown is used (even where KB isn’t relevant).

Potential fix
Keep the dropdown generic, and inject the “create” UI from outside:
Add a prop like renderCreateModal (or CreateModalComponent) / onCreateRequested.
Or make a wrapper component like KnowledgeBaseDropdown that composes the generic dropdown + KB modal logic.

File: src/frontend/src/components/core/parameterRenderComponent/components/modelInputComponent/index.tsx [hunk 3]
```diff
@@ -288,7 +294,7 @@ export default function ModelInputComponent({
       <div className="flex w-full flex-col">
         <PopoverTrigger asChild>
           <Button
-            disabled={disabled || options.length === 0}
+            disabled={disabled || (options.length === 0 && !showEmptyState)}
```
Comment (olayinkaadelakun) [inline src/frontend/src/components/core/parameterRenderComponent/components/modelInputComponent/index.tsx]: This logic is complex, can we move it all to the disable variable. Or just create one variable to store it all

File: src/frontend/src/modals/knowledgeBaseUploadModal/components/columnConfig/index.tsx [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+export function ColumnConfig({
+  columnConfig,
+  onColumnConfigChange,
+}: ColumnConfigProps) {
+  const AgColumns = FormatColumns(COLUMN_CONFIG_COLUMNS);
+  const agGrid = useRef<AgGridReact>(null);
+  const [isTableModalOpen, setIsTableModalOpen] = useState(false);
+  const [tempColumnConfig, setTempColumnConfig] = useState<ColumnConfigRow[]>(
+    cloneDeep(columnConfig),
```
Comment (olayinkaadelakun) [inline src/frontend/src/modals/knowledgeBaseUploadModal/components/columnConfig/index.tsx]: Why do we need to cloneDeep here. I believe a regular shallowCopy should be fine here 
```
function cloneColumnConfig(rows: ColumnConfigRow[]): ColumnConfigRow[] {
  return rows.map((row) => ({ ...row }));
}
```

That way we reduce the overhead of using a function like cloneDeep

**Comment by olayinkaadelakun** [comment]
Outstanding work @deon-sanchez. Super frontend heavy and executed well

**Comment by olayinkaadelakun** [comment]
I notices two issues while testing 
1. Flaky Icon (It has shown since then. But it would be nice to investigate)
Screenshot 2026-02-23 at 11 16 20 AM

2. When **updating** the knowledge base with new files, the preview doesn't seem to work
Screenshot 2026-02-23 at 11 21 45 AM

3. 
4. **Reach** We are currently just uploading the file when it is a duplicate. Should be doing anything to futher restrict this flow
5.  When uploading files, pictures aren't allowed, But if i upload a folder with images, it seems to upload.
6. I am also seeing Hidden files being uploaded like .DSStore
7. We have to delete each uploaded file individually, maybe considering a bulk delete functionality 
8. Screenshot 2026-02-23 at 12 02 17 PM
9.  Bulk delete of knowledge base item, shift|+click not selecting multiple items

File: src/backend/base/langflow/alembic/versions/169b35510b37_added_job_type_to_make_jobs_generic_and_.py (hunk 1 above)
Comment (dkaushik94) [reply src/backend/base/langflow/alembic/versions/169b35510b37_added_job_type_to_make_jobs_generic_and_.py]: This is a key check, makes the migration more stable. I will add this change. Good callout @ogabrielluiz !

File: src/frontend/src/components/core/dropdownComponent/index.tsx (hunk 2 above)
Comment (deon-sanchez) [reply src/frontend/src/components/core/dropdownComponent/index.tsx:631]: i agree with this.

File: src/frontend/src/modals/knowledgeBaseUploadModal/components/columnConfig/index.tsx (hunk 4 above)
Comment (deon-sanchez) [reply src/frontend/src/modals/knowledgeBaseUploadModal/components/columnConfig/index.tsx]: agreed

File: src/frontend/src/components/core/parameterRenderComponent/components/modelInputComponent/index.tsx (hunk 3 above)
Comment (deon-sanchez) [reply src/frontend/src/components/core/parameterRenderComponent/components/modelInputComponent/index.tsx]: ill skip this. if we go 4 logic deep, ill move it. but 3 isnt so bad.

File: src/backend/base/langflow/api/utils/kb_helpers.py [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+    total_words = 0
+    total_characters = 0
+
+    for col in text_columns:
+        if col not in df.columns:
+            continue
+
+        text_series = df[col].astype(str).fillna("")
+        total_characters += int(text_series.str.len().sum())
+        total_words += int(text_series.str.split().str.len().sum())
```
[the remark this answers is not in the document]
Comment (dkaushik94) [reply src/backend/base/langflow/api/utils/kb_helpers.py]: Addressed the review comments.

**Comment by Cristhianzl** [comment]
## Code Review - Remaining Items
### IMPORTANT — Must fix before merge

#### 1. Internal errors exposed to end users (`knowledge_bases.py`)

7 instances of `{e!s}` leaking internal details (paths, SQLite errors, class names) to the client:

```
Line 239:  detail=f"Error previewing chunks: {e!s}"
Line 363:  detail=f"Error ingesting files to knowledge base: {e!s}"
Line 460:  detail=f"Error listing knowledge bases: {e!s}"
Line 501:  detail=f"Error getting knowledge base '{kb_name}': {e!s}"
Line 587:  detail=f"Error getting chunks for '{kb_name}': {e!s}"
Line 608:  detail=f"Error deleting knowledge base '{kb_name}': {e!s}"
Line 655:  detail=f"Error deleting knowledge bases: {e!s}"
Line 712:  detail=f"Error cancelling ingestion: {e!s}"
```

**Fix:** Log the full error internally, return a generic message:
```python
logger.exception("Error creating knowledge base")
raise HTTPException(status_code=500, detail="Internal error creating knowledge base")
```

---

#### 2. Silent failure on `column_config` parsing (`knowledge_bases.py:296-297`)

```python
except (json.JSONDecodeError, TypeError):
    pass  # Ignore malformed column_config; use existing schema
```

Malformed user data is silently swallowed with no logging. This can cause unexpected behavior that is very hard to debug.

**Fix:** Add at minimum a `logger.warning`:
```python
except (json.JSONDecodeError, TypeError) as e:
    logger.warning("Malformed column_config received, using existing schema: %s", e)
```

---

#### 3. Broad `except Exception` with `noqa: BLE001` (7 occurrences)

`knowledge_bases.py` (2):
- Line 82
- Line 226

`kb_helpers.py` (5):
- Lines 69, 93, 147, 181, 447

These suppress the linter instead of handling specific exceptions. Most can be narrowed to `OSError`, `chromadb.errors.ChromaError`, `json.JSONDecodeError`, etc.

---

### RECOMMENDED — Should fix

#### 4. f-strings in logger calls

Multiple files use f-string interpolation in logger calls, which causes unnecessary string allocation even when the log level is disabled:

```python
# Current
logger.warning(f"Initial Chroma setup for {kb_name} failed: {e}")

# Preferred
logger.warning("Initial Chroma setup for %s failed: %s", kb_name, e)
```

#### 5. Redundant comments

`# Helper methods moved to utils/kb_helpers.py` appears multiple times — version control already tracks this.

---

### TESTING — Gaps

#### 6. Weak assertion in `test_perform_ingestion_rollback`

```python
# test_knowledge_bases_api.py line 487
assert mock_build is not None  # This is not a real assertion
```

This only verifies the mock object exists, not that it was called or returned expected values.

#### 7. No test for malformed `column_config`

The `except (json.JSONDecodeError, TypeError): pass` at line 296 has no test coverage. A test sending malformed JSON in `column_config` should verify the endpoint still succeeds using existing schema.

---

### Summary

| Category | Count | Status |
|----------|-------|--------|
| CRITICAL (Blockers) | 0 | All resolved |
| IMPORTANT (Must fix) | **3** | Items 1, 2, 3 |
| RECOMMENDED | 2 | Items 4, 5 |
| TESTING gaps | 2 | Items 6, 7 |

Resolving the 3 IMPORTANT items unblocks approval.
